apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+                    _hardStopCts.Cancel();
+                    ShutdownIfCompleted();
+                });
+                _taskHelper.Run(async () =>
+                {
+                    await _taskHelper.ConfigureAwaitHideErrors(_shutdownTcs.Task);
+                    registration.Dispose();
+                });
+                ShutdownIfCompleted();
+            }
+            return _shutdownTcs.Task;
+        }
+
+        private void CheckShutdown() => GaxPreconditions.CheckState(!_softStopCts.IsCancellationRequested, "Publisher already shutdown, cannot use");
+
+        private void ShutdownIfCompleted()
+        {
+            // Pre-condition: Must be locked if not a hard stop
+            if (_hardStopCts.IsCancellationRequested || (_softStopCts.IsCancellationRequested && _batchesInFlightCount == 0 && _batchesReady.Count == 0))
+            {
+                if (_shutdownLock.Locked(() => {
+                    var shutdownStarted = _shutdownStarted;
+                    _shutdownStarted = true;
+                    return !shutdownStarted;
+                }))
+                {
+                    // All batches sent and shutdown requested, so signal shutdown completed successfully.
+                    _taskHelper.Run(async () =>
+                    {
+                        if (_shutdown != null)
+                        {
+                            await _taskHelper.ConfigureAwaitHideErrors(_shutdown());
+                        }
+                        if (_hardStopCts.IsCancellationRequested)
+                        {
+                            _shutdownTcs.SetCanceled();
+                        }
+                        else
+                        {
+                            _shutdownTcs.SetResult(0);
+                        }
+                    });
+                }
+            }
+        }
+
+        // Not "async void" to gain standard async exeption handling
+        private async Task CreateBatch()
+        {
+            // Pre-condition: Must be locked - lock active up until first 'await'.
+            _currentBatch = new Batch();
+            var timerToken = _currentBatch.TimerCts.Token;
+            // After time-delay threshold, send the batch regardless of fullness.
+            await _taskHelper.ConfigureAwait(_scheduler.Delay(_batchDelayThreshold, timerToken));
+            // If batch has already moved to queue, timerToken will have been cancelled.
+            lock (_lock)
+            {
+                // Check for cancellation inside lock to avoid race-condition.
+                if (!timerToken.IsCancellationRequested)
+                {
+                    // Force queuing of the current batch, whatever the size.
+                    // There will always be at least one message in the batch.
+                    QueueCurrentBatchIfRequired(true, 0);
+                }
+            }
+        }
+
+        private void QueueCurrentBatchIfRequired(bool force, int extraByteCount)
```

Comment:
I think I'd split this into two methods:

- QueueCurrentBatchIfByteCountExceeded (or similar)
- QueueCurrentBatch

The former could call the latter...